keps/sig-multicluster/2149-clusterid/README.md

```diff
@@ -0,0 +1,766 @@
+and make progress.
+-->
+* Define any characteristics of the system that tracks cluster IDs within a cluster (i.e. a cluster registry)
+* Solve any problems without specific, tangible use cases (though we will leave room for extension).
+  * In particular, this KEP does not currently consider a cluster joining multiple ClusterSets
+
+
+## Proposal
+
+<!--
+This is where we get down to the specifics of what the proposal actually is.
+This should have enough detail that reviewers can understand exactly what
+you're proposing, but should not include things like API designs or
+implementation. What is the desired outcome and how do we measure success?.
+The "Design Details" section below is for the real
+nitty-gritty.
+-->
+
+### Overview
+Each cluster in a ClusterSet will be assigned a unique identifier, that lives at least as long as that cluster is a member of the given ClusterSet, and is immutable for that same lifetime. This identifier will be stored in a new cluster-scoped `ClusterClaim` CR with the well known name `id.k8s.io` that may be referenced by workloads within the cluster. The identifier must be a valid [RFC-1123](https://tools.ietf.org/html/rfc1123) DNS label, and may be created by an implementation dependent mechanism.
+
+While a member of a ClusterSet, a cluster will also have an additional `clusterset.k8s.io ClusterClaim` which describes its current membership. This claim must be present exactly as long as the cluster's membership in a ClusterSet lasts, and removed when the cluster is no longer a member.
+
+More detail and examples of the uniqueness, lifespan, immutability, and content requirements for both the `id.k8s.io ClusterClaim` and `clusterset.k8s.io ClusterClaim` are described further below. The goal of these requirements are to provide to the MCS API a cluster ID of viable usefulness to address known user stories without being too restrictive or prescriptive.
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### ClusterSet membership
+
+I have some set of clusters working together and need a way to uniquely identify them within the system that I use to track membership.
+
+#### Multi-Cluster Services
+
+I have a headless multi-cluster service deployed across clusters in my ClusterSet with similarly named pods in each cluster. I need a way to disambiguate each backend pod via DNS.
+
+```
+<<[UNRESOLVED]>>
+Examples of DNS using cluster ID a la `<hostname>.<clusterID>.<svc>.<ns>.svc.clusterset.local`
+<<[/UNRESOLVED]>>
+```
+
+#### Diagnostics
+
+Clusters within my ClusterSet send logs/metrics to a common monitoring solution and I need to be able to identify the cluster from which a given set of events originated.
+
+
+#### Joining or moving between ClusterSets
```

Comment:
I think this one should be next to clusterset membership above.

Reply:
Fixed in 1b782f7